File: src/main/java/de/vedantwankha/kata/dsa/LinkedList.java
package de.vedantwankha.kata.dsa;

import java.util.EmptyStackException;
import java.util.Iterator;

public class LinkedList<E> extends AbstractCollection<E> implements List<E>, ImmutableList<E>, Stack<E>, Deque<E> {
    private static class Node<E> {
        private E data;
        private Node<E> next;
        private Node<E> prev;

        Node(E e) {
            data = e;
            next = null;
            prev = null;
        }
    }

    private Node<E> head;
    private Node<E> tail;
    private int size;

    public LinkedList() {
        head = null;
        tail = null;
        size = 0;
    }

    public LinkedList(Collection<? extends E> c) {
        this();
        for (E e: c) {
            this.add(e);
        }
    }

    @Override
    public void add(E e) {
        addLast(e);
    }

    @Override
    public E popLast() {
        E e = get(size() - 1);
        tail = tail.prev;
        size--;
        return e;
    }

    @Override
    public void pushFirst(E e) {
        push(e);
    }

    @Override
    public void pushLast(E e) {
        addLast(e);
    }

    @Override
    public E popFirst() {
        return pop();
    }

    @Override
    public void add(E e, int idx) {
        if (idx > size) {
            throw new IndexOutOfBoundsException("Index " + idx + " should be less than or equal to the size of list " + size);
        } else if (idx == size) {
            addLast(e);
            return;
        } else if (idx == 0) {
            addFirst(e);
            return;
        }

        var currentNode = head;
        int i = 0;
        while (currentNode.next != null && i < idx - 1) {
            i++;
            currentNode = currentNode.next;
        }
        var newNode = new Node<E>(e);
        newNode.prev = currentNode.prev;
        newNode.next = currentNode.next;
        currentNode.prev = newNode;
        currentNode.next = newNode;
        size++;
    }

    @Override
    public void addFirst(E e) {
        var newNode = new Node<E>(e);
        if (head == null) {
            head = newNode;
            tail = newNode;
        } else {
            newNode.next = head;
            head.prev = newNode;
            head = newNode;
        }
        size++;
    }

    @Override
    public void addLast(E e) {
        var node = new Node<E>(e);
        if (head == null) {
            head = node;
            tail = node;
        } else {
            tail.next = node;
            node.prev = tail;
            tail = node;
        }
        size++;
    }

    @Override
    public void clear() {
        head = tail = null;
        size = 0;
    }

    @Override
    public E get(int idx) {
        if (head == null) throw new IndexOutOfBoundsException("List is empty");
        if (idx >= size) throw new IndexOutOfBoundsException("Given index " + idx + " should be less than size of the list " + size);
        int i = 0;
        var currentNode = head;
        while (currentNode.next != null && i < idx) {
            i++;
            currentNode = currentNode.next;
        }
        if (i == idx) return currentNode.data;
        else throw new IndexOutOfBoundsException("Given index " + idx + " should be less than size of the list " + size);
    }

    @Override
    public E getFirst() {
        return this.get(0);
    }

    @Override
    public E getLast() {
        return this.get(this.size - 1);
    }

    @Override
    public void remove(int idx) {
        if (idx >= size()) throw new IndexOutOfBoundsException("Given index " + idx + " should be less than the size of list " + size());
        if (idx == 0) {
            head = head.next;
            size--;
            return;
        }
        var currentNode = head;
        int i = 0;
        while (currentNode.next != null && i < idx - 1) {
            i++;
            currentNode = currentNode.next;
        }
        assert currentNode.next != null;
        currentNode.next = currentNode.next.next;
        size--;
    }

    @Override
    public void remove(E item) {
        // remove first occurrence
        if (head == null) throw new RuntimeException("Empty List");
        if (head.data.equals(item)) {
            head = head.next;
        } else {
            var current = head;
            while (current.next != null) {
                if (current.next.data.equals(item)) {
                    var toRemove = current.next;
                    current.next = toRemove.next;
                    if (toRemove.next != null) {
                        toRemove.next.prev = current;
                    } else {
                        tail = current;
                    }
                    size--;
                    break;
                }
                current = current.next;
            }
        }
    }

    @Override
    public void set(int i, E e) {
        // TODO))
    }

    @Override
    public int size() {
        return size;
    }

    @Override
    public boolean isEmpty() {
        return size == 0;
    }

    @Override
    public E pop() {
        if (size() == 0) throw new EmptyStackException();
        E e = get(0);
        remove(0);
        return e;
    }

    @Override
    public void push(E e) {
        addFirst(e);
    }

    @Override
    public Iterator<E> iterator() {
        return new LinkedListIterator();
    }

    private class LinkedListIterator implements Iterator<E> {

        private Node<E> currentNode = head;
        private int cursor = 0;

        @Override
        public boolean hasNext() {
            if (currentNode == null)
                return false;
            return true;
        }

        @Override
        public E next() {
            var ret = currentNode.data;
            currentNode = currentNode.next;
            cursor++;
            return ret;
        }
    }
}

File: sequence/linkedList.go
package sequence

import (
	"bytes"
	"fmt"
	"iter"
)

type node[T any] struct {
	data T
	next *node[T]
	prev *node[T]
}

type linkedList[T any] struct {
	head   *node[T]
	tail   *node[T]
	length int
}

func NewLinkedList[T any]() *linkedList[T] {
	return &linkedList[T]{
		head:   nil,
		tail:   nil,
		length: 0,
	}
}

func (ll *linkedList[T]) Len() int {
	return ll.length
}

func (ll *linkedList[T]) Add(t T) {
	newNode := &node[T]{
		data: t,
		next: nil,
	}
	if ll.head == nil {
		newNode.prev = nil
		ll.head = newNode
		ll.tail = newNode
	} else {
		newNode.prev = ll.tail
		ll.tail.next = newNode
		ll.tail = newNode
	}
	ll.length++
}

func (ll *linkedList[T]) AddAt(index int, t T) bool {
	if index < 0 || index > ll.Len() {
		return false
	}
	if index == ll.Len() {
		ll.Add(t)
		return true
	}
	if index == 0 {
		newNode := &node[T]{
			data: t,
			next: ll.head,
			prev: nil,
		}
		ll.head = newNode
		ll.length++
		return true
	}
	var n *node[T] = ll.head
	for i := 0; i < index; i, n = i+1, n.next {
	}
	newNode := &node[T]{
		data: t,
		next: n,
		prev: n.prev,
	}
	n.prev.next = newNode
	n.prev = newNode
	ll.length++
	return true
}

func (ll *linkedList[T]) Get(index int) (T, bool) {
	if index >= ll.Len() {
		return *new(T), false
	}

	n := ll.head

	for i := 0; i < index; i++ {
		n = n.next
	}

	return n.data, true
}

func (ll *linkedList[T]) String() string {
	var out bytes.Buffer
	out.WriteString("[")
	for n := ll.head; n != nil; n = n.next {
		out.WriteString(fmt.Sprintf("%v, ", n.data))
	}
	out.WriteString("]")
	return out.String()
}

func (ll *linkedList[T]) Set(index int, t T) bool {
	if index < 0 || index >= ll.Len() {
		return false
	}
	n := ll.head
	for i := 0; i < index; i, n = i+1, n.next {

	}
	n.data = t
	return true
}

func (ll *linkedList[T]) Iter() iter.Seq2[int, T] {
	return func(yield func(int, T) bool) {
		for i, n := 0, ll.head; n != nil; i, n = i+1, n.next {
			if !yield(i, n.data) {
				return
			}
		}
	}
}
